nixos/modules/services/web-apps/mastodon.nix

```diff
@@ -587,27 +601,166 @@ in {
 
     services.nginx = lib.mkIf cfg.configureNginx {
       enable = true;
-      recommendedProxySettings = true; # required for redirections to work
+
+      commonHttpConfig = ''
+        proxy_cache_path /var/cache/nginx levels=1:2 keys_zone=all_cache:10m inactive=7d max_size=1g;
+      '';
+
       virtualHosts."${cfg.localDomain}" = {
         root = "${cfg.package}/public/";
         forceSSL = true; # mastodon only supports https
         enableACME = true;
 
-        locations."/system/".alias = "/var/lib/mastodon/public-system/";
-
         locations."/" = {
           tryFiles = "$uri @proxy";
+          priority = 1100;
         };
 
-        locations."@proxy" = {
-          proxyPass = (if cfg.enableUnixSocket then "http://unix:/run/mastodon-web/web.socket" else "http://127.0.0.1:${toString(cfg.webPort)}");
-          proxyWebsockets = true;
+        locations."= /sw.js" = {
+          tryFiles = "$uri =404";
+          priority = 2110;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=604800, must-revalidate';
+            ${nginxCommonHeaders}
+          '';
+        };
+
+        locations."= /sw.js.map" = {
+          tryFiles = "$uri =404";
+          priority = 2120;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=604800, must-revalidate';
+            ${nginxCommonHeaders}
+          '';
+        };
+
+        locations."^~ /assets/" = {
+          tryFiles = "$uri =404";
+          priority = 2210;
+
+          extraConfig = ''
+            add_header Cache-Control 'public, max-age=2419200, must-revalidate';
+            ${nginxCommonHeaders}
+          '';
+        };
+
```

Comment:
There is a lot of repetition here. Can you use `listToAttrs` or so for `locations`?

Reply:
Duplicate lines have already been moved to the `nginxCommonHeaders` variable.

Reply:
there's still a lot of duplication, particularly the `tryFiles` and `Cache-Control` headers, which are the same for each location except sw.js. +1 to turion's suggestion of `listToAttrs`, especially if `priority` isn't necessary

Reply:
I don't see how it can be done. Are there any examples?

Reply:
in #202408 i declare `sidekiqUnits` which calls `lib.attrsets.mapAttrs'` to transform `cfg.sidekiqProcesses` (an attrset of attrsets containing `jobClasses` and `threads`) into a bunch of systemd units based on a template. I then merge those units with the rest of the config using `lib.attrsets.recursiveUpdate`.

I think you could take a similar approach here by declaring an attrset of attrsets containing things like the cache-control max-age and the headers to use. you could also have an `extraAttrs` attr that you merge into the location's attrset and taking precedence over whatever's in the template, if you wanted maximum flexibility for setting location-specific settings, or you could have an `extraConfig` attr that is simply appended.

Reply:
I think this will make it harder to read nginx configuration.
I used a similar variant in PeerTube service:
https://github.com/NixOS/nixpkgs/blob/2afac7a20c199d8cbd2645597c02beb5edfa7614/nixos/modules/services/web-apps/peertube.nix#L467-L744